taichi/rhi/metal/metal_device.mm

```diff
@@ -0,0 +1,495 @@
+#include "taichi/rhi/metal/metal_device.h"
+#include "spirv_msl.hpp"
+#include "taichi/rhi/device.h"
+#include "taichi/rhi/device_capability.h"
+
+namespace taichi::lang {
+namespace metal {
+
+MetalMemory::MetalMemory(MTLBuffer_id mtl_buffer) : mtl_buffer_(mtl_buffer) {}
+MetalMemory::~MetalMemory() { [mtl_buffer_ release]; }
+
+MTLBuffer_id MetalMemory::mtl_buffer() const { return mtl_buffer_; }
+size_t MetalMemory::size() const { return (size_t)[mtl_buffer_ length]; }
+RhiResult MetalMemory::mapped_ptr(void **mapped_ptr) const {
+  void *ptr = [mtl_buffer_ contents];
+  if (ptr == nullptr) {
+    return RhiResult::invalid_usage;
+  } else {
+    *mapped_ptr = ptr;
+    return RhiResult::success;
+  }
+}
+
+MetalPipeline::MetalPipeline(
+    const MetalDevice &device, MTLLibrary_id mtl_library,
+    MTLFunction_id mtl_function,
+    MTLComputePipelineState_id mtl_compute_pipeline_state,
+    MetalWorkgroupSize workgroup_size)
+    : device_(&device), mtl_library_(mtl_library), mtl_function_(mtl_function),
+      mtl_compute_pipeline_state_(mtl_compute_pipeline_state),
+      workgroup_size_(workgroup_size) {}
+MetalPipeline::~MetalPipeline() { destroy(); }
+MetalPipeline *MetalPipeline::create(const MetalDevice &device,
+                                     const uint32_t *spv_data,
+                                     size_t spv_size) {
+  TI_ASSERT((size_t)spv_data % sizeof(uint32_t) == 0);
```

Comment:
Don't use TI_ASSERT here (or any TI_xxx macros in fact), throw a exception and catch it from the Device API interface (& convert into return values, but that would need to wait for the API update for create_pipeline)